testdata/p4_16_samples_outputs/issue1937-2-bmv2-first.p4

```diff
@@ -0,0 +1,59 @@
+#include <core.p4>
+#include <v1model.p4>
+
+header h1_t {
+    bit<8> f1;
+    bit<8> f2;
+}
+
+struct headers_t {
+    h1_t h1;
+}
+
+struct metadata_t {
+}
+
+parser foo(out bit<8> x, in bit<8> y=5) {
+    state start {
+        x = y >> 2;
+        transition accept;
+    }
+}
+
+parser parserImpl(packet_in packet, out headers_t hdr, inout metadata_t meta, inout standard_metadata_t stdmeta) {
+    @name("foo") foo() foo_inst;
+    @name("foo") foo() foo_inst_0;
```

Comment:
There are probably other occurrences of duplicate @name annotations in the intermediate compiler pass results, too, but I will only point out this one more explicitly.

Reply:
Hmm.  A closer reading on my part of the P4_16 language spec, version 1.1, Section 14.1 "Direct type invocation" seems to me to mean that this example program is illegal, according to this text:

"In the same scope, direct type invocation will result in a different local instance per invocation—however, instances of the same type will share the same global name, via the @name annotation. If the type contains controllable entities, then invoking it directly more than once in the same scope is illegal, because it will produce multiple controllable entities with the same fully-qualified control name."

If that interpretation is correct, this example program should cause the compiler to give an error because of the multiple "automatic instantiations" of parser "foo".  The same would hold for the example program that does multiple "automatic instantiations" of a control.

Reply:
Parser are not controllable entities. They do not expose any APIs to the control-plane.

Reply:
Parsers can contain value_set instantiations, which do have a control plane API and internal state.

Reply:
If you will instantiate such a parser twice you should get an error. But if you instantiate a parser without any stateful objects I don't see why you should get an error. The problem is that the value_set has two instances with the same name, not the parser that contains it.

Reply:
So the P4_16 language allows arbitrary extern objects to be instantiated within parsers, too.

Are you saying that only a back-end compiler that knows whether the extern object has a control plane API or not, is the only place that such a program should be determined to be legal or illegal?

Reply:
I understand wanting to make more programs legal, but it seems like a reasonable simplification to the compiler, to tighten things up here and say it is always illegal to use multiple direct type invocations on the same parser or control name, regardless of whether it contains a control plane API or not.

Reply:
The code that generates the control-plane API should complain if it doesn't like something.
The fact that there is an extern does not mean that the extern is controllable by the control-plane using its name. After inlining these two parser instances will disappear anyway, so by the time you reach the back-end there aren't any more objects named the same way.